sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala

```diff
@@ -90,56 +88,69 @@ object SerializedHostTableUtils {
   }
 }
 
+/**
+ * Class that is used to broadcast results (a contiguous host batch) to executors.
+ *
+ * This is instantiated in the driver, serialized to an output stream provided by Spark
+ * to broadcast, and deserialized on the executor. Both the driver's and executor's copies
+ * are cleaned via GC. Because Spark closes `AutoCloseable` broadcast results after spilling
+ * to disk, this class does not subclass `AutoCloseable`. Instead we implement a `closeInternal`
+ * method only to be triggered via GC.
+ *
+ * @param data HostConcatResult populated for a broadcast that has column, otherwise it is null.
+ *             It is transient because we want the executor to deserialize its `data` from Spark's
+ *             torrent-backed input stream.
+ * @param output used to find the schema for this broadcast batch
+ * @param numRows number of rows for this broadcast batch
+ * @param dataLen size in bytes for this broadcast batch
+ */
 // scalastyle:off no.finalize
 @SerialVersionUID(100L)
 class SerializeConcatHostBuffersDeserializeBatch(
```

Comment:
It is not about the name, but that is part of it. It is the fact that we have multiple different Serializable classes that take one type of input and produce a different type of output. Why? The original reason was because we needed to broadcast an object from the host and put it on the device, where it would be deduped on the device. But is that really saving us anything. Would it be better to just send/receive a HostBuffer and the headers.  Just one Serializable class that does that. Then everything else are helper functions that transform it in different ways.

Not that this blocks the current patch from going in. I just thought that was the direction we were going to try, but I guess I was wrong.

Reply:
I started to file a follow on but I am not entirely sure how I'd merge the two serializables into one, well I can think of one way. Here's what happens today:

1. Every partition produces a `SerializeBatchDeserializeHostBuffer` from the executors. This class has to be serializable takes a `ColumnarBatch` and it is able to write and read from the streams.
2. The driver collects `Array[SerializeBatchDeserializeHostBuffer]`, where each element is a partition.
3. The driver concats this array and produces `HostConcatResult`.
4. The driver wraps the host concatenated result using `SerializeConcatHostBuffersDeserializeBatch` and broadcasts this out. 
5. Executors deal with the `HostConcatResult` only.

In order to remove the number of serializable classes, I think we could broadcast the Array that we collected in (2) as `Array` is `Serializable` and then the executors would concatenate these results when materializing the batch:

1. Every partition produces a `SerializeBatchDeserializeHostBuffer` from the executors. This class has to be serializable takes a `ColumnarBatch` and it is able to write and read from the streams.
2. The driver collects `Array[SerializeBatchDeserializeHostBuffer]`, where each element is a partition.
3. The driver broadcasts the `Array` directly.
4. Executors deal with serializing and deserializing this Array upto materialization. 
5. At materialiation, the executors create a `HostConcatResult` and turn it into a spillable GPU batch.

Reply:
ok I synced with @revans2 here and the main comment here was that the driver could still concatenate but then broadcast out essentially the `SerializeBatchDeserializeHostBuffer` again (as the result of the concatenation is a header and a host buffer, so this object could be reused). This would help with simplifying the code and make it easier for us to go to the next step to improve Broadcast which was to allow these broadcast GPU `RapidsBuffer` to be freed instead of actually spilled, since they are already in host memory at the source. We can work on this cleanup as part of https://github.com/NVIDIA/spark-rapids/issues/7709.